src/components/Switch/Switch.tsx

```diff
@@ -1,184 +1,231 @@
- *   </figure>
- *   <figure>
- *     <img src="screenshots/switch-disabled.android.png" />
- *     <figcaption>Android (disabled)</figcaption>
- *   </figure>
- *   <figure>
- *     <img src="screenshots/switch-enabled.ios.png" />
- *     <figcaption>iOS (enabled)</figcaption>
- *   </figure>
- *   <figure>
- *     <img src="screenshots/switch-disabled.ios.png" />
- *     <figcaption>iOS (disabled)</figcaption>
- *   </figure>
- * </div>
- *
- * ## Usage
- * ```js
- * import * as React from 'react';
- * import { Switch } from 'react-native-paper';
- *
- * const MyComponent = () => {
- *   const [isSwitchOn, setIsSwitchOn] = React.useState(false);
- *
- *   const onToggleSwitch = () => setIsSwitchOn(!isSwitchOn);
- *
- *   return <Switch value={isSwitchOn} onValueChange={onToggleSwitch} />;
- * };
- *
- * export default MyComponent;
- * ```
- */
-const Switch = (
-    {
-        value,
-        disabled,
-        onValueChange,
-        color,
-        style,
-        onTintColor: onTintColorProp,
-        thumbTintColor: thumbTintColorProp,
-        ...rest
-    }: Props,
-    ref: any,
-) => {
-    const componentStyles = useComponentStyles(
-        'Switch',
-        [
-            style,
-            {
-                checkedColor: color,
-                onTintColor: onTintColorProp,
-                thumbTintColor: thumbTintColorProp,
-            },
-        ],
-        {
-            states: {
-                selected_disabled: !!value && !!disabled,
-                active: !!value,
-                disabled: !!disabled,
-            },
+    const componentStyles = useComponentStyles('Switch', style, {
+        states: {
+            selected_hovered: !!value && !!hovered,
+            selected_disabled: !!value && !!disabled,
+            selected_pressed: !!value && !!pressed,
+            selected_focused: !!value && !!focused,
+            selected: !!value,
+            disabled: !!disabled,
+            hovered: !!hovered,
+            focused: !!focused,
+            pressed: !!pressed,
         },
-    );
+    });
+    const toggleMarginAnimation = useRef(new Animated.Value(value ? 0 : 1)).current;
+    const toggleSizeAnimation = useRef(new Animated.Value(value ? 0 : 1)).current;
+
+    const moveToggle = toggleMarginAnimation.interpolate({
+        inputRange: [0, 1],
+        outputRange: [
+            !value && pressed && !unCheckedIcon
+                ? 0
+                : !value && unCheckedIcon
+                ? size * 0.04
+                : size * 0.2,
+            pressed ? size * 0.74 : size * 0.81,
+        ],
+    });
 
-    const { switchStyle, props } = useMemo(() => {
-        const {
-            checkedColor,
-            onTintColor: _onTintColor,
-            thumbTintColor: _thumbTintColor,
-            ...switchStyles
-        } = componentStyles;
+    const toggleSize = toggleSizeAnimation.interpolate({
+        inputRange: [0, 1],
+        outputRange: [
```

Comment:
This is confusing. I think we can simplify this.

Reply:
These conditions are necessary, we can't simplify it more. @thetaungg

Reply:
We can, you just have to create consts and put them outside of this function